src/Components/DataHistoryExplorer/DataHistoryExplorer.tsx

```diff
@@ -39,24 +80,371 @@ const DataHistoryExplorer: React.FC<IDataHistoryExplorerProps> = (props) => {
+            }
+        });
+    }, [dispatch, state.explorerChartOptions.defaultQuickTimeSpanInMillis]);
+
+    const handleTimeSeriesTwinCalloutPrimaryAction = useCallback(
+        (timeSeriesTwin: IDataHistoryTimeSeriesTwin) => {
+            if (isDefined(state.selectedTimeSeriesId)) {
+                const selectedIdx = state.timeSeriesTwins.findIndex(
+                    (tsTwin) => tsTwin.seriesId === state.selectedTimeSeriesId
+                );
+                dispatch({
+                    type: DataHistoryExplorerActionType.EDIT_TIME_SERIES_TWINS,
+                    payload: { seriesIdx: selectedIdx, series: timeSeriesTwin }
+                });
+                const telemetry =
+                    TelemetryEvents.Tools.DataHistoryExplorer.UserAction
+                        .EditSeries;
+                sendDataHistoryExplorerUserTelemetry(telemetry.eventName, [
+                    {
+                        [telemetry.properties.itemIndex]: selectedIdx
+                    },
+                    {
+                        [telemetry.properties.hasCustomLabel]:
+                            timeSeriesTwin.label !==
+                            getDefaultSeriesLabel(
+                                timeSeriesTwin.twinId,
+                                timeSeriesTwin.twinPropertyName
+                            )
+                    }
+                ]);
+            } else {
+                dispatch({
+                    type: DataHistoryExplorerActionType.ADD_TIME_SERIES_TWINS,
+                    payload: {
+                        series: {
+                            ...timeSeriesTwin,
+                            chartProps: {
+                                ...timeSeriesTwin.chartProps,
+                                color: getHighChartColor(
+                                    usedSeriesColorsRef.current
+                                )
+                            }
+                        }
+                    }
+                });
+                const telemetry =
+                    TelemetryEvents.Tools.DataHistoryExplorer.UserAction
+                        .AddSeries;
+                sendDataHistoryExplorerUserTelemetry(telemetry.eventName, [
+                    {
+                        [telemetry.properties.hasCustomLabel]:
+                            timeSeriesTwin.label !==
+                            getDefaultSeriesLabel(
+                                timeSeriesTwin.twinId,
+                                timeSeriesTwin.twinPropertyName
+                            )
+                    }
+                ]);
+            }
+        },
+        [
+            isDefined,
+            dispatch,
+            sendDataHistoryExplorerUserTelemetry,
+            getDefaultSeriesLabel,
+            getHighChartColor,
+            state.timeSeriesTwins,
+            state.selectedTimeSeriesId
+        ]
+    );
+
+    const handleOnAddSeriesClick = useCallback(
```

Comment:
great stuff. love what you did here with the callbacks and reducing the dependency on the context in the lower components. Great idea.